src/util/bitset.h

```diff
@@ -0,0 +1,510 @@
+    /** Return an object with the binary AND between respective bits from a and b. */
+    friend constexpr IntBitSet operator&(const IntBitSet& a, const IntBitSet& b) noexcept { return I(a.m_val & b.m_val); }
+    /** Return an object with the binary OR between respective bits from a and b. */
+    friend constexpr IntBitSet operator|(const IntBitSet& a, const IntBitSet& b) noexcept { return I(a.m_val | b.m_val); }
+    /** Return an object with the binary AND NOT between respective bits from a and b. */
+    friend constexpr IntBitSet operator-(const IntBitSet& a, const IntBitSet& b) noexcept { return I(a.m_val & ~b.m_val); }
+    /** Return an object with the binary XOR between respective bits from a and b. */
+    friend constexpr IntBitSet operator^(const IntBitSet& a, const IntBitSet& b) noexcept { return I(a.m_val ^ b.m_val); }
+    /** Check if bitset a and bitset b are identical. */
+    friend constexpr bool operator==(const IntBitSet& a, const IntBitSet& b) noexcept = default;
+    /** Check if bitset a is a superset of bitset b (= every 1 bit in b is also in a). */
+    constexpr bool IsSupersetOf(const IntBitSet& a) const noexcept { return (a.m_val & ~m_val) == 0; }
+    /** Check if bitset a is a subset of bitset b (= every 1 bit in a is also in b). */
+    constexpr bool IsSubsetOf(const IntBitSet& a) const noexcept { return (m_val & ~a.m_val) == 0; }
+    /** Swap two bitsets. */
+    friend constexpr void swap(IntBitSet& a, IntBitSet& b) noexcept { std::swap(a.m_val, b.m_val); }
+};
+
+/** A bitset implementation backed by N integers of type I. */
+template<typename I, unsigned N>
+class MultiIntBitSet
+{
+    // Only binary, unsigned, integer, types allowed.
+    static_assert(std::is_integral_v<I> && std::is_unsigned_v<I> && std::numeric_limits<I>::radix == 2);
+    /** The number of bits per integer. */
+    static constexpr unsigned LIMB_BITS = std::numeric_limits<I>::digits;
+    /** Number of elements this set type supports. */
+    static constexpr unsigned MAX_SIZE = LIMB_BITS * N;
+    /** Array whose member integers store the bits of the set. */
+    std::array<I, N> m_val;
+    /** Dummy type to return using end(). Only used for comparing with Iterator. */
+    class IteratorEnd
+    {
+        friend class MultiIntBitSet;
+        constexpr IteratorEnd() = default;
+    public:
+        constexpr IteratorEnd(const IteratorEnd&) = default;
+    };
+    /** Iterator type returned by begin(), which efficiently iterates all 1 positions. */
+    class Iterator
+    {
+        friend class MultiIntBitSet;
+        const std::array<I, N>* m_ptr; /**< Pointer to array to fetch bits from. */
+        I m_val; /**< The remaining bits of (*m_ptr)[m_idx]. */
+        unsigned m_pos; /**< The last reported position. */
+        unsigned m_idx; /**< The index in *m_ptr currently being iterated over. */
+        constexpr Iterator(const std::array<I, N>* ptr) noexcept : m_ptr(ptr), m_idx(0)
```

Comment:
nit: `ptr` could be passed as a reference to signify non-nullness. (But the member still needs to be a pointer). Not at all significant as it is a private function.